src/frontend/flight/sync/utils.rs

```diff
@@ -414,6 +415,206 @@ pub(super) fn get_prune_map(
+                } else if upper_schema
+                    .column(lower_sync_col.name(), ColumnRole::Value)
+                    .is_some()
+                {
+                    // The upper schema doesn't have the `Changed` column at hand, but it does have
+                    // the corresponding `Value` column, project true to pick those values up
+                    lit(true)
+                } else {
+                    // The upper sync has neither the `Changed` nor the associated `Value` column,
+                    // project false to avoid overriding NULL values
+                    lit(false)
+                }
+            }
+            ColumnRole::Value => {
+                if maybe_upper_sync_col.is_none()
+                    && lower_schema
+                        .column(lower_sync_col.name(), ColumnRole::Changed)
+                        .is_none()
+                {
+                    // There's no corresponding `Value` sync column in the upper schema, and
+                    // there's no `Changed` column in the lower schema either, we need to project
+                    // the `Changed` column that takes all the values from the lower schema and
+                    // no value from the upper one
+                    col_desc.insert(ColumnDescriptor {
+                        role: ColumnRole::Changed as _,
+                        name: lower_sync_col.name().clone(),
+                    });
+                    projection.push(
+                        when(col(LOWER_SYNC).is_null(), lit(false))
+                            .otherwise(lit(true))?,
+                    );
+                }
+
+                if let Some(upper_value_col) = maybe_upper_sync_col {
+                    col(Column::new(
+                        Some(UPPER_SYNC),
+                        upper_value_col.field().name(),
+                    ))
+                } else {
+                    // The upper sync doesn't have this column but the lower sync already has
+                    // the `Changed` column which was/will be included in the output.
+                    // Project NULLs in its place.
+                    when(
+                        col(UPPER_SYNC).is_null(),
+                        lit(ScalarValue::Null
+                            .cast_to(lower_sync_col.field().data_type())?),
+                    )
+                    .otherwise(col(Column::new(
+                        Some(LOWER_SYNC),
+                        lower_sync_col.field().name(),
+                    )))?
+                }
+            }
+        };
+
+        col_desc.insert(ColumnDescriptor {
+            role: lower_sync_col.role() as _,
+            name: lower_sync_col.name().clone(),
+        });
+        projection.push(
+            when(
+                col(UPPER_SYNC).is_null(),
+                col(Column::new(Some(LOWER_SYNC), lower_sync_col.field().name())),
+            )
+            .otherwise(otherwise)?,
+        );
+    }
```

Comment:
This whole loop can be modularized like this. Which can be reused in the loop below for the `upper_sync_col`.  (Please verify if I'm not breaking the code). 

```suggestion
    for lower_sync_col in lower_schema.columns() {
        let expr = match lower_sync_col.role() {
            ColumnRole::OldPk | ColumnRole::NewPk => project_pk_columns(lower_sync_col, upper_schema),
            ColumnRole::Changed => project_changed_columns(lower_sync_col, upper_schema),
            ColumnRole::Value => project_value_columns(lower_sync_col, lower_schema, upper_schema),
        }?;
        projection.push(expr);

        col_desc.insert(ColumnDescriptor {
            role: lower_sync_col.role() as _,
            name: lower_sync_col.name().clone(),
        });
    }
```

With: 

```
fn project_pk_columns(
    sync_col: &SyncColumn,
    upper_schema: &SyncSchema,
) -> SyncResult<Expr> {
    let upper_sync_col = upper_schema
        .column(sync_col.name(), sync_col.role())
        .expect("Old/New Pk columns must be the same across syncs");

    Ok(when(
        col(LOWER_SYNC).is_null(),
        col(Column::new(Some(UPPER_SYNC), upper_sync_col.field().name())),
    )
    .otherwise(col(Column::new(Some(LOWER_SYNC), sync_col.field().name())))?)
}

fn project_changed_columns(
    sync_col: &SyncColumn,
    other_schema: &SyncSchema,
) -> SyncResult<Expr> {
    let upper_changed_col = other_schema.column(sync_col.name(), sync_col.role());

    Ok(if upper_changed_col.is_some() {
        col(Column::new(Some(UPPER_SYNC), upper_changed_col.field().name()))
    } else if other_schema.column(sync_col.name(), ColumnRole::Value).is_some() {
        lit(true)
    } else {
        lit(false)
    })
}

fn project_value_columns(
    sync_col: &SyncColumn,
    lower_schema: &SyncSchema,
    upper_schema: &SyncSchema,
) -> SyncResult<Expr> {
    let upper_value_col = upper_schema.column(sync_col.name(), ColumnRole::Value);

    Ok(if upper_value_col.is_some() {
        col(Column::new(Some(UPPER_SYNC), upper_value_col.field().name()))
    } else {
        when(
            col(UPPER_SYNC).is_null(),
            lit(ScalarValue::Null.cast_to(sync_col.field().data_type())?),
        )
        .otherwise(col(Column::new(Some(LOWER_SYNC), sync_col.field().name())))?
    })
}
```

Reply:
Yeah, I was looking into doing something like that, but there are some (non-)subtle differences between lower and upper sync column handling, which precluded them from sharing the same function. Or at least it would necessitate adding a bunch of extra parameters for instance the `WHEN` clause match expr, relation qualifier, the second `THEN` clause expr (the one in `otherwise`) etc.

In addition, the  `project_value_columns` would need to have `mut projection` and `mut col_desc` since there's a case when it needs to add an extra column that is missing in both upper and lower schema.

Let me see if I can simplify it somehow.

Reply:
Ok I refactored it a bit, should be a bit more readable now.

Reply:
Ok, merged `project_lower_changed_column` and `project_upper_changed_column` into `project_changed_column`, I think this helps too.

Will try compressing Value projection too.

Reply:
With the lower and upper functions is already much more readable. Thank you for the changes